controllers/rsyncnodedatamovement_controller.go

```diff
@@ -106,63 +162,106 @@ func (r *RsyncNodeDataMovementReconciler) Reconcile(ctx context.Context, req ctr
-		cmd.SysProcAttr = &syscall.SysProcAttr{
-			Credential: &syscall.Credential{
-				Uid: rsyncNode.Spec.UserId,
-				Gid: rsyncNode.Spec.GroupId,
-			},
+		// Set the appropriate permissions
+		if rsyncNode.Spec.UserId != uint32(os.Getuid()) || rsyncNode.Spec.GroupId != uint32(os.Getgid()) {
+			cmd.SysProcAttr = &syscall.SysProcAttr{
+				Credential: &syscall.Credential{
+					Uid: rsyncNode.Spec.UserId,
+					Gid: rsyncNode.Spec.GroupId,
+				},
+			}
 		}
-	}
-	log.Info("Rsync credential", "uid", rsyncNode.Spec.UserId, "gid", rsyncNode.Spec.GroupId)
+		log.Info("Rsync credential", "uid", rsyncNode.Spec.UserId, "gid", rsyncNode.Spec.GroupId)
 
-	out, err := cmd.Output()
+		// Run the command and wait for completion
+		log.Info("Executing rsync command", "source", source, "destination", destination)
+		out, cmdErr := cmd.CombinedOutput()
 
-	if err != nil {
-		if exitErr, ok := err.(*exec.ExitError); ok {
-			log.Info("Rsync failure", "error", string(exitErr.Stderr))
+		// Log output based on result
+		if errors.Is(ctxCancel.Err(), context.Canceled) {
+			log.Info("Rsync cancelled", "error", cmdErr, "output", string(out))
+		} else if cmdErr != nil {
+			log.Info("Rsync failure", "error", cmdErr, "output", string(out))
 		} else {
-			log.Info("Rsync failure", "error", err, "output", string(out))
+			log.Info("rsync completed", "output", string(out))
 		}
-	} else {
-		log.Info("rsync completed", "output", string(out))
-	}
 
-	// Record the completion status.
-	rsyncNode.Status.EndTime = metav1.NowMicro()
-	rsyncNode.Status.State = nnfv1alpha1.DataMovementConditionTypeFinished
+		// Record the completion status and retry on conflicts
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			rsyncNode = &dmv1alpha1.RsyncNodeDataMovement{}
+			if err := r.Get(ctx, req.NamespacedName, rsyncNode); err != nil {
+				return err
+			}
 
-	if err != nil {
-		rsyncNode.Status.Status = nnfv1alpha1.DataMovementConditionReasonFailed
-		rsyncNode.Status.Message = err.Error()
-	} else {
-		rsyncNode.Status.Status = nnfv1alpha1.DataMovementConditionReasonSuccess
-	}
+			rsyncNode.Status.EndTime = metav1.NowMicro()
+			rsyncNode.Status.State = nnfv1alpha1.DataMovementConditionTypeFinished
 
-	if err := r.Status().Update(ctx, rsyncNode); err != nil {
-		log.Error(err, "failed to update rsync status with completion")
-		r.recordCompletion(*rsyncNode)
-		return ctrl.Result{}, err
+			if errors.Is(ctxCancel.Err(), context.Canceled) {
+				rsyncNode.Status.Status = nnfv1alpha1.DataMovementConditionReasonCancelled
+			} else if cmdErr != nil {
+				rsyncNode.Status.Status = nnfv1alpha1.DataMovementConditionReasonFailed
+				rsyncNode.Status.Message = cmdErr.Error()
+			} else {
+				rsyncNode.Status.Status = nnfv1alpha1.DataMovementConditionReasonSuccess
+			}
+
+			// Record the completion in case the update does not work
+			r.recordCompletion(*rsyncNode)
```

Comment:
You only need to do this _if_ the update fails for resource conflict after you've exhausted retries
```
	if errors.IsConflict(err) {

		log.Info("Status update conflicted")

		// Record the completion internally so when we requeue we won't
		// restart the operation but update the resource with saved status.
		r.completions.Store(req.Name, dm.Status)
```

Reply:
The reason I put this here is because we have access to `rsyncNode` which contains the `EndTime`, `State`, `Status` that we're trying to update. If we take this out of RetryOnConflict then we lose that information. That was my thought process in putting it there and then removing it after the update is successful (on line 249).

Reply:
You could move the variable out of the RetryOnConflict

Reply:
Moving the variable out means we need to move the `Get()` out of the `RetryOnConflict`. How susceptible to failure is the `Get()`?

If it does fail, then we have an empty `rsyncNode` and we won't retry the `Get()`. In that case, if we try to `recordCompletion` then we'll lose the state of the rsync since we don't have any of the information and this goroutine will be lost.

I'm having trouble seeing a good way to handle the `Get()` failure outside of `RetryOnConflict`.

Reply:
If `errors.IsConflict(err)` is true, you know the `Get()` worked and it's just the status couldn't be updated; so we'll try again next Reconcile

Reply:
The Get() has to be inside the RetryOnConflict.  Otherwise, as RetryOnConflict is looping it'll never refresh its version of the resource, and it'll always get a conflict.

Reply:
This has been reverted. The `Get()` has been moved back inside the `RetryOnConflict` block.